servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/HedgingDnsNameResolver.java

```diff
@@ -0,0 +1,289 @@
+        return setupHedge(delegate::resolveAll, t);
+    }
+
+    @Override
+    public void close() {
+        delegate.close();
+    }
+
+    private long currentTimeMillis() {
+        return executor.currentTime(TimeUnit.MILLISECONDS);
+    }
+
+    private <T, R> Future<R> setupHedge(Function<T, Future<R>> computation, T t) {
+        // Only add tokens for organic requests and not retries.
+        budget.deposit();
+        Future<R> underlyingResult = computation.apply(t);
+        final long delay = percentile.getValue();
+        if (delay == Long.MAX_VALUE) {
+            // basically forever: just return the value.
+            return underlyingResult;
+        } else {
+            final long startTimeMs = currentTimeMillis();
+            Promise<R> promise = executor.eventLoopGroup().next().newPromise();
+            Cancellable hedgeTimer = executor.schedule(() -> tryHedge(computation, t, underlyingResult, promise),
+                    delay, TimeUnit.MILLISECONDS);
+            underlyingResult.addListener(completedFuture -> {
+                measureRequest(currentTimeMillis() - startTimeMs, completedFuture);
+                if (complete(underlyingResult, promise)) {
+                    hedgeTimer.cancel();
+                }
+            });
+            return promise;
+        }
+    }
+
+    private <T, R> void tryHedge(
+            Function<T, Future<R>> computation, T t, Future<R> original, Promise<R> promise) {
+        if (!original.isDone() && budget.withdraw()) {
+            System.out.println("" + System.currentTimeMillis() + ": sending backup request.");
+            Future<R> backupResult = computation.apply(t);
+            final long startTime = currentTimeMillis();
+            backupResult.addListener(done -> {
+                if (complete(backupResult, promise)) {
+                    original.cancel(true);
+                    measureRequest(currentTimeMillis() - startTime, done);
+                }
+            });
+            promise.addListener(complete -> backupResult.cancel(true));
+        }
+    }
+
+    private void measureRequest(long durationMs, Future<?> future) {
+        // Cancelled responses don't count but we do consider failed responses because failure
+        // is a legitimate response.
+        if (!future.isCancelled()) {
+            percentile.addSample(durationMs);
+        }
+    }
+
+    private <T, R> boolean complete(Future<R> f, Promise<R> p) {
+        assert f.isDone();
+        if (f.isSuccess()) {
+            return p.trySuccess(f.getNow());
+        } else {
+            return p.tryFailure(f.cause());
+        }
+    }
+
+    interface PercentileTracker {
+        void addSample(long sample);
+
+        long getValue();
+    }
+
+    interface Budget {
+        void deposit();
+
+        boolean withdraw();
+    }
+
+    // TODO: both these implementations are un-synchronized and rely on netty using only a single event loop.
```

Comment:
We always use only a single event loop and this is ensured by `inEventLoop` checks across `DefaultDnsClient`